### ISSUE TITLE
@IsOptional with @IsNotEmpty is not working

### ISSUE THREAD

**Issue by vsternbach** [typestack/class-validator#491]
This code:
```typescript
import { IsNotEmpty, IsOptional, IsString, validate } from 'class-validator';
import { plainToClass } from 'class-transformer';

const PATCH = 'patch';
const POST = 'post';

export class Test {
  @IsOptional({ groups: [PATCH] })
  @IsNotEmpty({ always: true })
  @IsString()
  name: string;
}

async function getValidationErrors(obj, group) {
  return await validate(plainToClass(Test, obj), { groups: [group] });
}

describe('Test', () => {
  it('should fail on post without name', async () => {
    const errors = await getValidationErrors({}, POST);
    expect(errors).not.toEqual([]);
  });
  it('should fail on post when name is undefined', async () => {
    const errors = await getValidationErrors({ name: undefined }, POST);
    expect(errors).not.toEqual([]);
  });
  it('should fail on post when name is null', async () => {
    const errors = await getValidationErrors({ name: null }, POST);
    expect(errors).not.toEqual([]);
  });
  it('should fail on post when name is empty', async () => {
    const errors = await getValidationErrors({ name: '' }, POST);
    expect(errors).not.toEqual([]);
  });
  it('should succeed on patch without name property', async () => {
    const errors = await getValidationErrors({}, PATCH);
    expect(errors).toEqual([]);
  });
  it('should fail on patch when name is undefined', async () => {
    const errors = await getValidationErrors({ name: undefined }, PATCH);
    expect(errors).not.toEqual([]);
  });
  it('should fail on patch when name is null', async () => {
    const errors = await getValidationErrors({ name: null }, PATCH);
    expect(errors).not.toEqual([]);
  });
  it('should fail on patch when name is empty', async () => {
    const errors = await getValidationErrors({ name: '' }, PATCH);
    expect(errors).not.toEqual([]);
  });
});
```
gives the following results:
<img width="363" alt="Screen Shot 2019-12-19 at 18 15 43" src="https://user-images.githubusercontent.com/3830668/71189511-9c854680-228b-11ea-9d0a-9d047cd44258.png">
Am I missing something here?

Looking at the code of @IsOptional I see that it checks that the property is not undefined or null, that's why the test for empty string is not failing, but in theory shouldn't it check that this property exists on the object?

**Comment by vlapo** [comment]
I do not understand your problem. I think your test cases are not good. You define property `name` as optional in case PATCH is in groups. 
```ts
it('should fail on patch when name is undefined', async () => {
    const errors = await getValidationErrors({ name: undefined }, PATCH);
    expect(errors).not.toEqual([]);
  });
  it('should fail on patch when name is null', async () => {
    const errors = await getValidationErrors({ name: null }, PATCH);
    expect(errors).not.toEqual([]);
  });
```
So `Test should succeed on patch when name is undefined/null`.
From docs:

@IsOptional() | Checks if given value is empty (=== null, === undefined) and if so, ignores all the validators on the property.
-- | --

**Comment by Kiyozz** [comment]
I use nestjs to validate a DTO from my controller

---
I have this

```
class MyDTO {
  @IsOptional()
  @IsNotEmpty()
  locale?: string
}
```

I send this body json

```
{
  "locale": null
}
```

The expected behaviour:
- ERROR 400, locale should not be empty

The actual behaviour:
- CODE 200, locale `null` is ignored.

---
If I have this DTO

```
class MyDTO {
  @IsNotEmpty()
  locale?: string
}
```

Then `null` value throws CODE 400 which is expected.

---


`@IsOptional()` should ignores only when the key is not passed in the request (aka. `undefined`).

**Comment by vlapo** [comment]
I understand now. But as it is a breaking change I want to have more feedback from community.

**Comment by jrcss11-iscteiulpt** [comment]
I don't know if is the same issue, but i need to validate if a variable is notEmpty, notNull and notUndefined. Is that possible?

**Comment by sam3d** [comment]
I've had to create my own decorator `IsOptionalStrict` which I've now realised is pretty much just this PR. The `IsOptional` decorator really should not ignore `null`, as it is a valid property that is very different from `undefined`.

**Comment by sam3d** [comment]
How about something like an extra configurator? This wouldn't break backwards compatibility:

https://github.com/se-internal/class-validator/commit/5ac9fa0c17ad43ecb0e8642675447b380a54139e

**Comment by fhp** [comment]
We would love this as well, ether with a nullable flag, or as a separate set of decorators (although naming them can be difficult).
@sam3d, your fork looks good to me.

**Comment by sam3d** [comment]
Thanks @fhp! Yeah we had originally tried to create a new decorator and must've spent over an hour trying to come up with a good name. Eventually, we decided it made more sense to borrow the design pattern from typeorm's  `@ManyToOne()` decorator (a `nullable` property that defaults to `true`).

I'll definitely put it into a PR at some point in the next couple of days! Just waiting until I have a moment to add tests. Confirmation from @vlapo or another maintainer that at least the design proposal is accepted would be amazing so I know I'm not doing it for nothing 😄

**Comment by omidh28** [comment]
Any update for this? really want to see this feature in next release 
@sam3d

**Comment by NoNameProvided** [comment]
I agree about the strange behavior of this. 

I like the API proposed by @sam3d. We can use the current behavior as default and add a warning in the changelog about eventually being flipped.

I will take a crack at this at the weekend.

**Comment by omidh28** [comment]
@NoNameProvided Any update? 10 days has been passed

**Comment by NickKelly1** [comment]
How about keeping `@IsOptional` as-is and adding new decorators `@IsNullable` and `@IsUndefinable`.
Any support for this?

**Comment by ruscon** [comment]
```typescript
/**
 * Skips validation if the target is null
 *
 * @example
 * ```typescript
 * class TestModel {
 *     @IsNullable({}, { always: true })
 *     big: string | null;
 * }
 * ```
 */
export function IsNullable(options?: ValidationOptions): PropertyDecorator {
    return function IsNullableDecorator(prototype: object, propertyKey: string | symbol): void {
        ValidateIf((obj): boolean => null !== obj[propertyKey], options)(prototype, propertyKey);
    };
}
```


```typescript
/**
 * Skips validation if the target is undefined
 *
 * @example
 * ```typescript
 * class TestModel {
 *     @IsUndefinable({}, { always: true })
 *     big?: string;
 * }
 * ```
 */
export function IsUndefinable(options?: ValidationOptions): PropertyDecorator {
    return function IsUndefinableDecorator(prototype: object, propertyKey: string | symbol): void {
        ValidateIf((obj): boolean => undefined !== obj[propertyKey], options)(prototype, propertyKey);
    };
}
```

**Comment by hfhchan-plb** [comment]
For other parts of TypeScript, "optional" means either the property is not set, or it is set to `undefined`. In no other places does "optional" imply "nullable", whereas "nullable" itself can sometimes imply "optional" (c.f. `NonNullable` utility type). The naming of `IsOptional` is really confusing.

I suggest the following changes to align with the terminology of TypeScript:
1. `@IsOptional()` taking a new `strict` parameter, defaulting to `false` which allows `null` for back-compat, and `true` for denying `null`. The `strict` parameter should default to `true` at some later version and possibly throw if it is set to `false`.
2. `@IsNullable` with a `strict` parameter defaulting to `true` which only allows `null`, and `false` for also allowing `undefined`.

**Comment by mwanago** [comment]
I think
```
@IsOptional({ nullable: false })
```
looks great, as suggested by @sam3d
@vlapo Are there any plans on implementing that?

**Comment by mwanago** [comment]
Meanwhile, I suggest creating two separate decorators:
```typescript
import { ValidateIf } from 'class-validator';

export function CanBeUndefined() {
  return ValidateIf((object, value) => value !== undefined);
}

export function CanBeNull() {
  return ValidateIf((object, value) => value !== null);
}
```

**Comment by pigulla** [comment]
In case anyone is interested, I've had the exact same issue and implemented this (alongside quite a few more decorators) in [class-validator-extended](https://github.com/pigulla/class-validator-extended/), specifically [`Optional()`](https://pigulla.github.io/class-validator-extended/functions/Optional.html) and [`Nullable()`](https://pigulla.github.io/class-validator-extended/functions/Nullable.html).

**Comment by oscarmeanwell** [comment]
This is still an issue, please  can this be updated? Some fantastic solutions have been presented. I am happy to make a PR

**Comment by fr1sk** [comment]
any update on this one? And what about `PartialType` which is making all props optional? Can we somehow pass this `nullable` param to `PartialType` as well? 🤔

**Comment by renatocron** [comment]
I also don't like this behavior, it a new function like IsOptionalNonNullable exists, and had integration with the nest/swagger plugin to also be automatically marked as optional in the swagger, would be great.

It leads to behaviors where a Prisma expect only a value or undef (like the typescript type said so) but if the client/frontend send null, leads the backend to error 500 with weird messages

**Comment by renatocron** [comment]
Hi @pigulla 

Is any helper like PartialType in class-validator-extended? I could not find it

Thank you

**Comment by pigulla** [comment]
@renatocron,

unfortunately not, I consider that out of scope for that library.

**Comment by pigulla** [comment]
@renatocron 

unfortunately not, I consider that out of scope for that library.
/

**Comment by oscarmeanwell** [comment]
I solved this myself and posted on [StackOverflow ](https://stackoverflow.com/a/76738269/8692914), I am happy to make a PR amending IsOptional or adding this to the library. 

```javascript
import { IsOptional, ValidateIf, ValidationOptions } from 'class-validator'

export function IsOptionalNonNullable(data?: {
  nullable: boolean
  validationOptions?: ValidationOptions
}) {
  const { nullable = false, validationOptions = undefined } = data || {}

  if (nullable) {
    // IsOptional allows null
    return IsOptional(validationOptions)
  }

  return ValidateIf((ob: any, v: any) => {
    return v !== undefined
  }, validationOptions)
}

// Example usage 
export class SomeUpdateDTO {
  @IsInt()
  // Param can be undefined, but not null 
  @IsOptionalNonNullable()
  nbOfViews?: number
}

// Example of why IsOptional is a problem
export class SomeOtherUpdateDTO {
  @IsInt()
  @IsOptional()
  // Null is not specified, but IsOptional will allow it! 
  // Could end up nulling a required field in the db
  nbOfViews?: number
}
```

**Comment by alexpizarroj** [comment]
@vlapo

I came across this exact problem today. I have an API with a field that **must** be present. The field represents an assigned user id, and I want it to support either an actual id string or `null` to clear the assignment. For this endpoint in particular, not receiving an `assignedUserId` field value at all (i.e., `undefined`) would be strange/ill-defined.

In the meantime, I'm going with the solution [here](https://github.com/typestack/class-validator/issues/491#issuecomment-693589520) to unblock myself, however I'd love to help in any way I can get something official merged. For example, we already have a great contribution in https://github.com/typestack/class-validator/pull/2443.

**Comment by alexpizarroj** [comment]
@vlapo @NoNameProvided 

I came across this exact problem today. I have an API with a field that **must** be present. The field represents an assigned user id, and I want it to support either an actual id string or `null` to clear the assignment. For this endpoint in particular, not receiving an `assignedUserId` field value at all (i.e., `undefined`) would be strange/ill-defined.

In the meantime, I'm going with the solution [here](https://github.com/typestack/class-validator/issues/491#issuecomment-693589520) to unblock myself, however I'd love to help in any way I can get something official merged. For example, we already have a great contribution in https://github.com/typestack/class-validator/pull/2443.

Thank you so much for all the work in this project 🙏🏻

**Comment by Jv7680** [comment]
I don't understand why they haven't had any updates on this issue for 5 years now.

But @IsOptional() accept null value is weird behavior and make me crazy when using NestJs PartialType.

**Comment by davidmwhynot** [comment]
It looks like there are several solutions for this issue ready to go... is there a timeline as to when at least one of them will be accepted? Seems like we're mostly just blocked by maintainers right now.

**Comment by braaar** [comment]
Is there an agreement about what the API should be like? I feel like every PR that attempts to fix this is slightly different from the other, since everyone has a slightly different idea about what undefined and null values mean to them

**Comment by braaar** [comment]
A classic example is what your REST API should do if it receives an undefined value in a PATCH call. E.g. an API call is made to update a user profile, and the bio field in the DTO is `undefined`. Should the value in the database be cleared, or rather left untouched? Some people want `undefined` to mean "don't touch this field" and `null` to mean "clear it", while others use actual values such as an empty string to clear values.

**Comment by braaar** [comment]
I think the new `validateIf` validator option should be helpful here: https://github.com/typestack/class-validator/pull/1579

**Comment by p-mcgowan** [comment]
FWIW I ran into this only recently. My expectation was that an optional field which was not explicitly allowed to be null should fail if null is provided, but it did not.

my solution:
```ts
/** optional when undefined only, not undefined or null */
export const IsNotUndefined = (options?: ValidationOptions): PropertyDecorator => {
  return (prototype: object, propertyKey: string | symbol) => {
    ValidateIf(
      (obj: unknown) => typeof obj === 'object' && obj !== null && typeof obj[propertyKey as keyof typeof obj] !== 'undefined',
      options,
    )(prototype, propertyKey);
  };
};

export const NullTransform = (options?: TransformOptions): PropertyDecorator =>
  Transform(({ value }) => {
    if (value === null) {
      return null;
    }
    if (value !== '') {
      return value as unknown;
    }
  }, options);

export class SomethingDto {
  /** non-nullable field - if it's set, it contains a value */
  @Expose()
  @IsNotUndefined()
  @IsString()
  @MinLength(1)
  @Matches(/\S/)
  name?: string;

  /** nullable field - can be omitted, but can also be null (will unset the field) */
  @Expose()
  @IsNotUndefined()
  @IsString()
  @NullTransform()
  @ValidateIf((o?: { ['description']?: null }) => o?.['description'] !== null)
  @MinLength(1)
  @Matches(/\S/)
  description?: string | null;
}

await client
  .patch('/something/ae5738e1-54ce-4897-bc0b-69fba6d4cab3')
  .send({ description: 'hello' } satisfies SomethingDto)
  .expect(200);

await client
  .patch('/something/ae5738e1-54ce-4897-bc0b-69fba6d4cab3')
  .send({ description: null } satisfies SomethingDto)
  .expect(200);

await client
  .patch('/something/ae5738e1-54ce-4897-bc0b-69fba6d4cab3')
  .send({ description: null, name: undefined } satisfies SomethingDto)
  .expect(200);

await client
  .patch('/something/ae5738e1-54ce-4897-bc0b-69fba6d4cab3')
  .send({ description: null } satisfies SomethingDto)
  .expect(200);

await client
  .patch('/something/ae5738e1-54ce-4897-bc0b-69fba6d4cab3')
  .send({ name: null })
  .expect(422);
